dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -1342,6 +1363,18 @@ void PageDirectory<Trait>::apply(PageEntriesEdit && edit, const WriteLimiterPtr
         {
             // Protected in write_lock
             max_page_id = std::max(max_page_id, Trait::PageIdTrait::getU64ID(r.page_id));
+            if constexpr (std::is_same_v<Trait, universal::PageDirectoryTrait>)
+            {
+                // TODO: the value is just used at restart, if it slows down the performance, we can avoid calculating it at run time.
```

Comment:
Can we only calculate this at restart by adding a new param?

Reply:
This code path is only used at run time. Restart related code is located in `PageDirectoryFactory`. We can just remove this code when we need at any time.